internal/component/azure/servicebus/subscription.go

```diff
@@ -168,28 +226,90 @@ func (s *Subscription) ReceiveAndBlock(handler HandlerFunc, lockRenewalInSec int
+		s.logger.Debugf("Received messages: %d; current active operations usage: %d/%d", l, len(s.activeOperationsChan), cap(s.activeOperationsChan))
+
+		skipProcessing := false
+		for _, msg := range msgs {
+			if err = s.addActiveMessage(msg); err != nil {
+				// If we cannot add the message then sequence number is not set, this must
+				// be a bug in the Azure Service Bus SDK so we will log the error and not
+				// handle the message. The message will eventually be retried until fixed.
+				s.logger.Errorf("Error adding message: %s", err.Error())
+				skipProcessing = true
+				break
+			}
+			s.logger.Debugf("Processing received message: %s", msg.MessageID)
+		}
 
-		if err = s.addActiveMessage(msg); err != nil {
-			// If we cannot add the message then sequence number is not set, this must
-			// be a bug in the Azure Service Bus SDK so we will log the error and not
-			// handle the message. The message will eventually be retried until fixed.
-			s.logger.Errorf("Error adding message: %s", err.Error())
-			<-s.activeMessagesChan
+		if skipProcessing {
+			<-s.activeOperationsChan
 			continue
 		}
 
-		s.logger.Debugf("Processing received message: %s", msg.MessageID)
-		s.handleAsync(s.ctx, msg, handler)
+		runHandlerFn := func(hctx context.Context) {
+			msg := msgs[0]
+
+			// Invoke the handler to process the message
+			_, err = handler(hctx, msgs)
+
+			// This context is used for the calls to service bus to finalize (i.e. complete/abandon) the message.
+			// If we fail to finalize the message, this message will eventually be reprocessed (at-least once delivery).
+			// This uses a background context in case ctx has been canceled already.
+			finalizeCtx, finalizeCancel := context.WithTimeout(context.Background(), s.timeout)
+			defer finalizeCancel()
+
+			if err != nil {
+				// Log the error only, as we're running asynchronously
+				s.logger.Errorf("App handler returned an error for message %s on %s: %s", msg.MessageID, s.entity, err)
+				s.AbandonMessage(finalizeCtx, msg)
+				return
+			}
+
+			s.CompleteMessage(finalizeCtx, msg)
+		}
+
+		bulkRunHandlerFunc := func(hctx context.Context) {
+			resps, err := handler(hctx, msgs)
+
+			// This context is used for the calls to service bus to finalize (i.e. complete/abandon) the message.
+			// If we fail to finalize the message, this message will eventually be reprocessed (at-least once delivery).
+			// This uses a background context in case ctx has been canceled already.
+			finalizeCtx, finalizeCancel := context.WithTimeout(context.Background(), s.timeout)
+			defer finalizeCancel()
+
+			if err != nil {
+				// Handle the error and mark messages accordingly.
+				// Note, the order of the responses match the order of the messages.
+				for i, resp := range resps {
+					if resp.Error != nil {
+						// Log the error only, as we're running asynchronously.
+						s.logger.Errorf("App handler returned an error for message %s on %s: %s", msgs[i].MessageID, s.entity, err)
```

Comment:
May be wrong, but shouldn't `err` be `resp.Error` here?

Reply:
Yup, thanks!